# 01. Import Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# 02. Import Data

In [2]:
# Turn project folder path into a string
path = r'C:\Users\marze\CF Projects\08-2024 Instacart Basket Analysis'

In [3]:
# Import prodcuts.csv data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'),index_col = False)

In [4]:
# Import orders_wrangled.csv data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Consistency Check

In [5]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# 04. Mixed-Type Data/New dataframe

In [6]:
# create a dataframe
df_test  = pd.DataFrame()

In [7]:
# create a mixed-type column
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# function for checking whether a dataframe contains any mixed-type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
# change all values to string
df_test['mix'] = df_test['mix'].astype('str')

# 05. Missing Values

### A. Finding Missing Values

In [11]:
# assign the function isnull() to the df_prods dataframe, then sum the result with the attached sum() function. The isnull() function is used to find missing observations
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# create subset to display the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### B. Addressing Missing Values

In [14]:
# compare size of df_prods with new subset
df_prods.shape

(49693, 5)

In [15]:
# create new subset without the missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
# check size of the new subset
df_prods_clean.shape

(49677, 5)

# 06. Duplicates

### A. Finding Duplicates

In [17]:
#look for full duplicates/ create subset containing only rows that are duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### B. Addressing Duplicates

In [19]:
# check size of the dataframe
df_prods_clean.shape

(49677, 5)

In [20]:
# create new dataframe that doesn't include duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
# check size of new dataframe
df_prods_clean_no_dups.shape

(49672, 5)

# 07. Exporting Changes

In [22]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5

##### 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [23]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

The max for the order_number is 100, which could mean that the orders are capped at 100. 'days_since_prior_order' has a maximum value of 30 which means that the counter for the order number could be reseting every month. The minimum value of the same column is 0 which could suggest that there a multiple orders placed on the same day. 

##### 3. Check for mixed-type data in your df_ords dataframe.

In [24]:
# function for checking whether a dataframe contains any mixed-type columns
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

##### 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

There are no mixed-type data in this dataframe.

In [25]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

##### 5. Run a check for missing values in your df_ords dataframe.

In [26]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The 'days_since_prior_order' is missing 206209 values. It could mean that these are first time orders and they have not been assigned a value yet. 

##### 6. Address the missing values using an appropriate method.


In [27]:
# check size of df_ords
df_ords.shape

(3421083, 7)

In [28]:
# create new subset without the missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [29]:
df_ords_nan.shape

(206209, 7)

In [30]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [31]:
# check how often order_number=1 in df_ords_nan
df_ords_nan['order_number'].value_counts()

order_number
1    206209
Name: count, dtype: int64

In [32]:
# check how often order_number=1 in df_ords
df_ords['order_number'].value_counts()

order_number
1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: count, Length: 100, dtype: int64

In [33]:
# Filter the dataframe where 'order_number' equals 1
df_first_orders = df_ords_nan[df_ords_nan['order_number'] == '1']

In [34]:
# Check if all of these have missing values in 'days_since_prior_order'
df_first_orders_missing = df_first_orders['days_since_prior_order'].isna().all()

In [35]:
# Print the result
print(df_first_orders_missing)

True


In [36]:
# Address the missing values by creating a column that flags order numbers as either True/False depending on if the 'order_number' = 1
df_ords['first_order'] = df_ords['days_since_prior_order'].isna()

In [37]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 8 columns]

In [38]:
# Re-check data to make sure totals are accurate
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
first_order                    0
dtype: int64

I've addressed the missing values by creating a new column 'first_order' that will return a value of True for all NaN values in the 'days_since_prior_order' column. This will tell us if this a first time order but it will not remove the observations in that column. If these rows have been removed, we would've missed potentially valuable information for our analysis.

##### 7. Run a check for duplicate values in your df_ords data.
##### 8. Address the duplicates using an appropriate method.

In [39]:
#look for full duplicates/ create subset containing only rows that are duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [40]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

In [41]:
df_ords_dups.shape

(0, 8)

There are no duplicates found in our dataframe.

##### 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

The cleaned df_prods has been exported in the exercise portion 07. Exporting Changes

In [42]:
# exporting df_ords dataframe as "orders_checked.csv"
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))